### ISSUE TITLE
jsii/java issue - CfnTag: argument type mismatch

### ISSUE THREAD

**Issue by sai-tirunagiri** [awslabs/aws-cdk#1717]
Hi, Today I upgraded the cdk libraries from 0.23.0 -> 0.24.1 and made the changes from cdk.Tag to cdk.CfnTag post this change upon doing cdk synth ended up getting an **argument type mismatch** error.

source code:
```
tag = CfnTag.builder().withKey("key_name").withValue("tag_value").build();
tagList.add(tag);

CfnVPCProps.builder().withTags(tagList).build();
```
any solution to solve argument type mismatch error.?

cdk pom version : 0.24.1
cdk version : 0.24.1

**Comment by eladb** [comment]
Seems like a bug in how tags passed into properties of L1s interact with that recently added `TagManager` object in Java. 

As a workaround, you could directly access the tag manager like this:

```java
CfnVPC vpc = new CfnVPC(this, "MyVpc", CfnVPCProps.builder()
    .withCidrBlock("0.0.0.0/0")
    .build());

vpc.getTags().setTag("Hello", "Value");
```

Copy @moofish32

**Comment by eladb** [comment]
Here's a jsii dump of the repro:

```
/Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home/bin/java -ea -Didea.test.cyclic.buffer.size=1048576 "-javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=55133:/Applications/IntelliJ IDEA CE.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath "/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit-rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit5-rt.jar:/Users/benisrae/playground/pg-113/target/test-classes:/Users/benisrae/playground/pg-113/target/classes:/Users/benisrae/.m2/repository/software/amazon/awscdk/cdk/0.24.0/cdk-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/cdk-cx-api/0.24.0/cdk-cx-api-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/jsii/jsii-runtime/0.7.14/jsii-runtime-0.7.14.jar:/Users/benisrae/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.9.5/jackson-core-2.9.5.jar:/Users/benisrae/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.9.5/jackson-databind-2.9.5.jar:/Users/benisrae/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.9.0/jackson-annotations-2.9.0.jar:/Users/benisrae/.m2/repository/com/google/guava/guava-io/r03/guava-io-r03.jar:/Users/benisrae/.m2/repository/com/google/guava/guava-annotations/r03/guava-annotations-r03.jar:/Users/benisrae/.m2/repository/com/google/guava/guava-base/r03/guava-base-r03.jar:/Users/benisrae/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/iam/0.24.0/iam-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/s3/0.24.0/s3-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/codepipeline-api/0.24.0/codepipeline-api-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/events/0.24.0/events-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/kms/0.24.0/kms-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/s3-notifications/0.24.0/s3-notifications-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/sns/0.24.0/sns-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/autoscaling-api/0.24.0/autoscaling-api-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/cloudwatch/0.24.0/cloudwatch-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/lambda/0.24.0/lambda-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/cdk-assets/0.24.0/cdk-assets-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/ec2/0.24.0/ec2-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/stepfunctions/0.24.0/stepfunctions-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/cloudtrail/0.24.0/cloudtrail-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/logs/0.24.0/logs-0.24.0.jar:/Users/benisrae/.m2/repository/software/amazon/awscdk/sqs/0.24.0/sqs-0.24.0.jar:/Users/benisrae/.m2/repository/junit/junit/4.12/junit-4.12.jar:/Users/benisrae/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar" com.intellij.rt.execution.junit.JUnitStarter -ideVersion5 -junit4 com.myorg.HelloStackTest,testStack

software.amazon.jsii.JsiiException: While synthesizing test/MyVpc: argument type mismatch
While synthesizing test/MyVpc: argument type mismatch
  --- resource created at ---
  at s._wrapSandboxCode (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:86829)
  at t.Kernel._wrapSandboxCode (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:95312)
  at t.Kernel._create (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:86808)
  at t.Kernel.create (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:81792)
  at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78503)
  at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
  at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
  at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
  at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
  at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
  at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
  at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
  at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
  at i.then.e (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78831)
  --- problem discovered at ---
    at i (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78066)
    at t.KernelHost.callbackHandler (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78149)
    at t.Kernel.kernel.n.Kernel.e [as callbackHandler] (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:77764)
    at Object.set [as tags] (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:88472)
    at CfnVPC.toCloudFormation (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-kernel-pdldjs/node_modules/@aws-cdk/cdk/lib/cloudformation/resource.js:157:38)
    at node.recordReference.resolve_1.findTokens (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-kernel-pdldjs/node_modules/@aws-cdk/cdk/lib/cloudformation/stack-element.js:86:80)
    at Object.findTokens (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-kernel-pdldjs/node_modules/@aws-cdk/cdk/lib/core/tokens/resolve.js:106:17)
    at CfnVPC.prepare (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-kernel-pdldjs/node_modules/@aws-cdk/cdk/lib/cloudformation/stack-element.js:86:52)
    at ConstructNode.prepareTree (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-kernel-pdldjs/node_modules/@aws-cdk/cdk/lib/core/construct.js:257:27)
    at t.Kernel._wrapSandboxCode (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:95331)
    at a._ensureSync (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:83767)
    at t.Kernel._ensureSync (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:94848)
    at t.Kernel.invoke (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:83712)
    at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78503)
    at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
    at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
    at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
    at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
    at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
    at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
    at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
    at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
    at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
    at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
    at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
    at t.KernelHost.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78957)
    at t.KernelHost.run (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78192)
    at e.processRequest (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78218)
    at i.then.e (/private/var/folders/n2/6v4_tbz97ws0h4bn5gbyvzb0m8vcjb/T/jsii-java-runtime1576292816050053151/jsii-runtime.js:1:78831)

	at software.amazon.jsii.JsiiRuntime.processErrorResponse(JsiiRuntime.java:118)
	at software.amazon.jsii.JsiiRuntime.requestResponse(JsiiRuntime.java:90)
	at software.amazon.jsii.JsiiRuntime.processCallbackResponse(JsiiRuntime.java:163)
	at software.amazon.jsii.JsiiRuntime.requestResponse(JsiiRuntime.java:95)
	at software.amazon.jsii.JsiiRuntime.processCallbackResponse(JsiiRuntime.java:163)
	at software.amazon.jsii.JsiiRuntime.requestResponse(JsiiRuntime.java:95)
	at software.amazon.jsii.JsiiClient.callMethod(JsiiClient.java:196)
	at software.amazon.jsii.JsiiObject.jsiiCall(JsiiObject.java:61)
	at software.amazon.awscdk.App.synthesizeStack(App.java:39)
	at com.myorg.HelloStackTest.testStack(HelloStackTest.java:24)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by moofish32** [comment]
@eladb  The actual tag property from code generation is like:
```ts
    /**
     * ``AWS::Lambda::Function.Tags``
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-tags
     */
    tags?: Array<cdk.CfnTag | cdk.Token> | cdk.Token;
```
The TagManager is obviously a different type. Then in `resource.ts` there is this piece: https://github.com/awslabs/aws-cdk/blob/v0.24.1/packages/%40aws-cdk/cdk/lib/cloudformation/resource.ts#L209-L212

`properties.tags` is set to what I expect to a valid structure (an array of CfnTag, but perhaps there is an error or num2str type of conversion?), but I am suspicious if we don't somewhere cross the wires on the type of `tags`. I haven't had a chance to dig deeper, but that's what I thought of when I saw this issue.

**Comment by moofish32** [comment]
@eladb another interesting thing about this bug: 

```java
//        CfnQueue cfn = new CfnQueue(this, "RawQueue", CfnQueueProps.builder().withTags(tags).build());
        CfnQueue cfn = new CfnQueue(this, "RawQueue");
        cfn.apply(new Tag("Yay", "Aspects"));
```
If you switch the commented out queue with the current queue the behavior changes. The queue with props throws the error.

**Comment by eladb** [comment]
Verified that this has been resolved in 0.28.0
